### PR TITLE
mcobbett unit testable cobra commands

### PR DESCRIPTION
**Pull request by techcobweb**
- Initial objectifying the root command
- Several commands converted into GalasaCommand classes
- Properties and Auth commands are now GalasaCommands
- NamespaceGet is now a GalasaCommand
- Runs Get is now a GalasaCommand
- Runs Prepare is now a GalasaCommand
- Runs Submit is now a GalasaCommand
- Runs Submit Local is now a GalasaCommand

### CODE REVIEW

**Comment by galasa-team** [comment]
Build failed, see http://localhost:8001/api/v1/namespaces/tekton-pipelines/services/tekton-dashboard:http/proxy/#/namespaces/galasa-build/pipelineruns/repo-cli-pr-183-lpxhz for details. If you are unable to do so, please contact a member of the Galasa team.

File: pkg/cmd/authLogin.go [hunk 1]
```diff
@@ -18,35 +18,64 @@ type AuthLoginCmdValues struct {
...
+}
+
+func (cmd *AuthLoginComamnd) GetValues() interface{} {
+	return cmd.values
+}
+
+// ------------------------------------------------------------------------------------------------
+// Private methods
+// ------------------------------------------------------------------------------------------------
+func (cmd *AuthLoginComamnd) init(factory Factory, authCommand GalasaCommand, rootCommand GalasaCommand) error {
```
Comment (Wyvinar) [inline pkg/cmd/authLogin.go:53]: This section strays from other conventions. Should there be a change to have a seperate function for building the cobra and galasa command?

**Comment by galasa-team** [comment]
Build successful

File: pkg/cmd/commandCollection.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+type CommandCollection interface {
+	// name - One of the COMMAND_NAME_* constants.
+	GetCommand(name string) GalasaCommand
+
+	GetRootCommand() GalasaCommand
+
+	Execute(args []string) error
+}
+
+type CommandCollectionImpl struct {
```
Comment (techcobweb) [inline pkg/cmd/commandCollection.go:23]: Should this be a lower-case label to limit visibility of the implementation ?

File: pkg/cmd/commandCollection.go [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+	}
+
+	if err == nil {
+		sanitiseCommandHelpDescriptions(rootCommand.GetCobraCommand())
+	}
+
+	return err
+}
+
+func (commands *CommandCollectionImpl) addAuthCommands(factory Factory, rootCommand GalasaCommand) error {
```
Comment (techcobweb) [inline pkg/cmd/commandCollection.go:150]: Lots of repetition. We need an add function which takes a function as a parameter.

File: pkg/cmd/commandCollection.go [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+	commands := new(CommandCollectionImpl)
+
+	err := commands.init(factory)
+
+	return commands, err
+}
+
+// The main entry point into the cmd package.
+func Execute(factory Factory, args []string) error {
```
Comment (techcobweb) [inline pkg/cmd/commandCollection.go:65]: Needs to be an object method

File: pkg/cmd/commandCollection.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+		if len(nameAndAliases) > 1 {
+			alias = nameAndAliases[1]
+		}
+
+		cobra.Flags().BoolP("help", "h", false, "Displays the options for the "+alias+" command.")
+	})
+}
+
+// TODO: Make this an object method.
+func setHelpFlagForAllCommands(command *cobra.Command, setHelpFlag func(*cobra.Command)) {
```
Comment (techcobweb) [inline pkg/cmd/commandCollection.go:344]: Separate pull request to refactor this.

File: pkg/cmd/finalWordHandler.go [hunk 6]
```diff
@@ -17,18 +17,26 @@ import (
...
 
 func NewRealFinalWordHandler() FinalWordHandler {
-	return new(RealFinalWordHandler)
+	handler := new(RealFinalWordHandler)
+	handler.init()
+	return handler
 }
 
-func (*RealFinalWordHandler) FinalWord(obj interface{}) {
+func (handler *RealFinalWordHandler) init() {
```
Comment (techcobweb) [inline pkg/cmd/finalWordHandler.go:33]: What's this doing ?

Comment (techcobweb) [reply pkg/cmd/finalWordHandler.go:33]: needs removing.

**Comment by galasa-team** [comment]
Build successful

File: pkg/cmd/finalWordHandlerMock.go [hunk 7]
```diff
@@ -13,7 +13,7 @@ func NewMockFinalWordHandler() FinalWordHandler {
 	return new(MockFinalWordHandler)
 }
 
-func (this *MockFinalWordHandler) FinalWord(o interface{}) {
+func (this *MockFinalWordHandler) FinalWord(rootCmd GalasaCommand, o interface{}) {
```
Comment (techcobweb) [inline pkg/cmd/finalWordHandlerMock.go:16]: o needs a better name

File: pkg/cmd/finalWordHandler.go [hunk 8]
```diff
@@ -17,18 +17,26 @@ import (
...
+}
+
+func (handler *RealFinalWordHandler) FinalWord(rootCmd GalasaCommand, obj interface{}) {
```
Comment (Wyvinar) [inline pkg/cmd/finalWordHandler.go:36]: maybe rename obj here to be more descriptive?

File: pkg/cmd/local.go [hunk 9]
```diff
@@ -9,20 +9,40 @@ import (
...
+}
+
+func NewLocalCommand(factory Factory, rootCommand GalasaCommand) (GalasaCommand, error) {
+	cmd := new(LocalCommand)
+	err := cmd.init(factory, rootCommand)
+	return cmd, err
+}
 
-	localCmd := &cobra.Command{
+func (cmd *LocalCommand) init(factory Factory, rootCommand GalasaCommand) error {
```
Comment (techcobweb) [inline pkg/cmd/local.go:22]: Missing separator comments.

File: pkg/cmd/project.go [hunk 10]
```diff
@@ -19,14 +29,20 @@ func createProjectCmd(factory Factory, parentCmd *cobra.Command, rootCmdValues *
 		Long:  "Creates and manipulates Galasa test project source code",
 	}
 
-	parentCmd.AddCommand(projectCmd)
+	cmd.cobraCommand = projectCmd
+	rootCmd.GetCobraCommand().AddCommand(projectCmd)
 
-	err = createProjectCmdChildren(factory, projectCmd, rootCmdValues)
+	return err
+}
```
Comment (techcobweb) [inline pkg/cmd/project.go:36]: missing comment separators

File: pkg/cmd/project.go [hunk 11]
```diff
@@ -9,7 +9,17 @@ import (
...
+	cobraCommand *cobra.Command
+}
+
+func NewProjectCmd(factory Factory, rootCmd GalasaCommand) (GalasaCommand, error) {
+	cmd := new(ProjectCommand)
+	err := cmd.init(factory, rootCmd)
+	return cmd, err
+}
+
+func (cmd *ProjectCommand) init(factory Factory, rootCmd GalasaCommand) error {
```
Comment (Wyvinar) [inline pkg/cmd/project.go:22]: reorder and add seperating comments here

File: pkg/cmd/projectCreate.go [hunk 12]
```diff
@@ -42,7 +42,34 @@ type ProjectCreateCmdValues struct {
...
+	cobraCommand *cobra.Command
+	values       *ProjectCreateCmdValues
+}
+
+func NewProjectCreateCmd(factory Factory, rootCmd GalasaCommand, projectCmd GalasaCommand) (GalasaCommand, error) {
+	var err error = nil
+
+	cmd := new(ProjectCreateCommand)
+	err = cmd.init(factory, rootCmd, projectCmd)
+
```
Comment (techcobweb) [inline pkg/cmd/projectCreate.go:55]: missing comment separators

File: pkg/cmd/projectCreate.go [hunk 13]
```diff
@@ -42,7 +42,34 @@ type ProjectCreateCmdValues struct {
 	isDevelopmentProjectCreate bool
 }
 
-func createProjectCreateCmd(factory Factory, parentCmd *cobra.Command, rootCmdValues *RootCmdValues) (*cobra.Command, error) {
+type ProjectCreateCommand struct {
+	cobraCommand *cobra.Command
+	values       *ProjectCreateCmdValues
+}
```
Comment (Wyvinar) [inline pkg/cmd/projectCreate.go:48]: add seperating comments reminder :)

File: pkg/cmd/properties.go [hunk 14]
```diff
@@ -16,41 +16,58 @@ type PropertiesCmdValues struct {
...
+}
+
+func (cmd *PropertiesCommand) GetValues() interface{} {
+	return cmd.values
+}
+
+// ------------------------------------------------------------------------------------------------
+// Private methods
+// ------------------------------------------------------------------------------------------------
+func (cmd *PropertiesCommand) init(factory Factory, rootCommand GalasaCommand) error {
```
Comment (Wyvinar) [inline pkg/cmd/properties.go:52]: Init function contains create cobra command function again

File: pkg/cmd/propertiesGet.go [hunk 15]
```diff
@@ -31,50 +31,87 @@ type PropertiesGetCmdValues struct {
...
+
+func (cmd *PropertiesGetComamnd) GetValues() interface{} {
+	return cmd.values
+}
+
+// ------------------------------------------------------------------------------------------------
+// Private methods
+// ------------------------------------------------------------------------------------------------
+
+func (cmd *PropertiesGetComamnd) init(factory Factory, propertiesCommand GalasaCommand, rootCommand GalasaCommand) error {
```
Comment (Wyvinar) [inline pkg/cmd/propertiesGet.go:68]: init function contains create cobra command

File: pkg/cmd/propertiesSet.go [hunk 16]
```diff
@@ -25,36 +25,70 @@ type PropertiesSetCmdValues struct {
...
+}
+
+func (cmd *PropertiesSetCommand) GetValues() interface{} {
+	return cmd.values
+}
+
+// ------------------------------------------------------------------------------------------------
+// Private methods
+// ------------------------------------------------------------------------------------------------
+func (cmd *PropertiesSetCommand) init(factory Factory, propertiesCommand GalasaCommand, rootCommand GalasaCommand) error {
```
Comment (Wyvinar) [inline pkg/cmd/propertiesSet.go:61]: init function contains create cobra command function

File: pkg/cmd/runs.go [hunk 17]
```diff
@@ -13,41 +13,47 @@ type RunsCmdValues struct {
...
+
+func (cmd *RunsCommand) GetCobraCommand() *cobra.Command {
+	return cmd.cobraCommand
+}
+
+func (cmd *RunsCommand) GetValues() interface{} {
+	return cmd.values
+}
+
+func (cmd *RunsCommand) init(factory Factory, rootCommand GalasaCommand) error {
```
Comment (Wyvinar) [inline pkg/cmd/runs.go:39]: init function contains create cobra command function

File: pkg/cmd/propertiesNamespaceGet_test.go [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+ */
+package cmd
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPropertiesNamespaceGetCommandInCommandCollectionHasName(t *testing.T) {
```
Comment (techcobweb) [inline pkg/cmd/propertiesNamespaceGet_test.go:14]: name of this test is suspect

File: pkg/cmd/propertiesNamespace_test.go [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+ */
+package cmd
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPropertiesNamespaceCommandInCommandCollectionHasName(t *testing.T) {
```
Comment (techcobweb) [inline pkg/cmd/propertiesNamespace_test.go:14]: name of this test is suspect - it's checking far more than just the name.

File: pkg/cmd/propertiesSet_test.go [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+ */
+package cmd
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPropertiesSetCommandInCommandCollectionHasName(t *testing.T) {
```
Comment (techcobweb) [inline pkg/cmd/propertiesSet_test.go:14]: name of the test

File: pkg/cmd/properties_test.go [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+ */
+package cmd
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPropertiesCommandInCommandCollectionHasName(t *testing.T) {
```
Comment (techcobweb) [inline pkg/cmd/properties_test.go:14]: test name is bad

File: pkg/cmd/runs.go [hunk 22]
```diff
@@ -13,41 +13,47 @@ type RunsCmdValues struct {
...
+}
+
+func (cmd *RunsCommand) GetCobraCommand() *cobra.Command {
+	return cmd.cobraCommand
+}
+
+func (cmd *RunsCommand) GetValues() interface{} {
+	return cmd.values
+}
+
```
Comment (techcobweb) [inline pkg/cmd/runs.go:38]: section/region comments to break it up ?

**Comment by galasa-team** [comment]
Build successful

File: pkg/cmd/commandCollection.go (hunk 4 above)
Comment (techcobweb) [reply pkg/cmd/commandCollection.go]: no it doesn't

File: pkg/cmd/authLogin.go (hunk 1 above)
Comment (techcobweb) [reply pkg/cmd/authLogin.go:53]: Fixed.

File: pkg/cmd/commandCollection.go (hunk 2 above)
Comment (techcobweb) [reply pkg/cmd/commandCollection.go:23]: yes it should

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

File: pkg/cmd/finalWordHandlerMock.go (hunk 7 above)
Comment (Wyvinar) [reply pkg/cmd/finalWordHandlerMock.go:16]: renamed to "errorToExtract"

File: pkg/cmd/runs.go (hunk 22 above)
Comment (Wyvinar) [reply pkg/cmd/runs.go:38]: doesn't seem to consistently work accross devices

File: pkg/cmd/commandCollection.go [hunk 23]
```diff
@@ -0,0 +1,351 @@
...
+		runsSubmitLocalCommand, err = NewRunsSubmitLocalCommand(factory, runsSubmitCommand, runsCommand, rootCommand)
+		if err == nil {
+			commands.commandMap[runsSubmitLocalCommand.GetName()] = runsSubmitLocalCommand
+		}
+	}
+
+	return err
+}
+
+// TODO: Make this an object method.
```
[the remark this answers is not in the document]
Comment (Wyvinar) [reply pkg/cmd/commandCollection.go:288]: todo in a seperate branch

File: pkg/cmd/propertiesNamespaceGet_test.go (hunk 18 above)
Comment (Wyvinar) [reply pkg/cmd/propertiesNamespaceGet_test.go:14]: deleted "HasName" from end of test name

File: pkg/cmd/propertiesNamespace_test.go (hunk 19 above)
Comment (Wyvinar) [reply pkg/cmd/propertiesNamespace_test.go:14]: deleted "HasName" from test name

**Comment by galasa-team** [comment]
Build successful

**Review by Wyvinar — APPROVED** [review]

**Comment by galasa-team** [comment]
Build successful

**Pull request merged**